src/Components/AuthDialog/AuthDialogContext.tsx

```diff
@@ -0,0 +1,73 @@
+import { AuthDialog, AuthDialogProps } from "Components/AuthDialog/AuthDialog"
+import { createContext, FC, useContext, useReducer } from "react"
+
+export const AUTH_DIALOG_MODES = ["Login", "SignUp", "ForgotPassword"] as const
+
+export type AuthDialogMode = typeof AUTH_DIALOG_MODES[number]
+
+const AuthDialogContext = createContext<{
+  dispatch: React.Dispatch<Action>
+  showAuthDialog(options: { mode: AuthDialogMode }): void
+  state: { mode: AuthDialogMode }
+  hideAuthDialog(): void
+}>({
+  dispatch: () => null,
+  showAuthDialog: () => null,
+  state: { mode: "SignUp" },
+  hideAuthDialog: () => null,
+})
+
+type State = {
+  mode: AuthDialogMode
+  isVisible: boolean
+}
+
+type Action =
+  | { type: "MODE"; payload: { mode: AuthDialogMode } }
+  | { type: "SHOW" }
+  | { type: "HIDE" }
+
+const reducer = (state: State, action: Action) => {
+  switch (action.type) {
+    case "MODE":
+      return { ...state, mode: action.payload.mode }
+    case "SHOW":
+      return { ...state, isVisible: true }
+    case "HIDE":
+      return { ...state, isVisible: false }
+    default:
+      return state
+  }
+}
+
+export const AuthDialogProvider: FC<Omit<AuthDialogProps, "onClose">> = ({
+  children,
+}) => {
+  const [state, dispatch] = useReducer(reducer, {
+    mode: "SignUp",
+    isVisible: false,
+  })
+
+  const hideAuthDialog = () => {
+    dispatch({ type: "HIDE" })
+  }
+
+  const showAuthDialog = ({ mode }: { mode: AuthDialogMode }) => {
+    dispatch({ type: "MODE", payload: { mode } })
+    dispatch({ type: "SHOW" })
+  }
+
+  return (
+    <AuthDialogContext.Provider
+      value={{ state, dispatch, showAuthDialog, hideAuthDialog }}
+    >
+      {children}
+
+      {state.isVisible && <AuthDialog onClose={hideAuthDialog} />}
```

Comment:
So I'm designing the usage API to be such that, there's a single instance of this modal that we show/hide. You import and use `const { showAuthDialog } = useAuthDialog()` and then can use `showAuthDialog({ ... })` with options for how it's displayed.